### PR TITLE
Store-gateway: disable index header verification

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
#### What this PR does

This PR disables verifying the index header upon loading it. Reading the whole index header can be resource consuming. Verification/checksumming reads the whole index header once in addition to the read to sample the index header. This change should effectively halve the time to load an index header since one of the reads will be skipped.

@pracucci suggested to make this configurable per-tenant. But the changes would involve changing 23 tests and 3 more production code usages. I'm not sure if the complexity is worth it at this point. I can extend the PR if reviewers think per-tenant configuration is necessary. IMO this configuration parameter depends on the hardware on which the store-gateway runs and the risk for all tenants is the same.

#### Which issue(s) this PR fixes or relates to

Related to #5046 

#### Checklist

- [ ] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 ## main / unreleased
 
+* [CHANGE] Store-gateway: skip verifying index header integrity upon loading. To enable verification set `blocks_storage.bucket_store.index_header.verify_on_load: true`.
```
Comment (pracucci) [inline CHANGELOG.md:5]:
```suggestion
### Grafana Mimir

* [CHANGE] Store-gateway: skip verifying index header integrity upon loading. To enable verification set `-blocks_storage.bucket_store.index_header.verify_on_load=true`.
```

File: pkg/storegateway/indexheader/header.go [hunk 2]
```diff
@@ -53,8 +53,10 @@ type Reader interface {
 
 type Config struct {
 	MaxIdleFileHandles uint `yaml:"max_idle_file_handles" category:"advanced"`
+	VerifyOnLoad       bool `yaml:"verify_on_load" category:"advanced"`
 }
 
 func (cfg *Config) RegisterFlagsWithPrefix(f *flag.FlagSet, prefix string) {
-	f.UintVar(&cfg.MaxIdleFileHandles, prefix+"max-idle-file-handles", 1, "Maximum number of idle file handles the store-gateway keeps open for each index-header file.")
+	f.UintVar(&cfg.MaxIdleFileHandles, prefix+"max-idle-file-handles", 1, "Maximum number of idle file handles the store-gateway keeps open for each index header file.")
+	f.BoolVar(&cfg.VerifyOnLoad, prefix+"verify-index-headers-on-load", false, "If true, verify the checksum of index headers upon loading them (either on startup or lazily when lazy loading is enabled). Setting to true helps detect disk corruption at the cost of slowing down index header loading.")
```
Comment (pracucci) [inline pkg/storegateway/indexheader/header.go:61]: Please make the CLI flag and YAML config option match.

```suggestion
	f.BoolVar(&cfg.VerifyOnLoad, prefix+"verify-on-load", false, "If true, verify the checksum of index headers upon loading them (either on startup or lazily when lazy loading is enabled). Setting to true helps detect disk corruption at the cost of slowing down index header loading.")
```

File: pkg/storegateway/indexheader/index/symbols.go [hunk 3]
```diff
@@ -41,8 +41,13 @@ type Symbols struct {
...
 // NewSymbols returns a Symbols object for symbol lookups.
-func NewSymbols(factory *streamencoding.DecbufFactory, version, offset int) (s *Symbols, err error) {
-	d := factory.NewDecbufAtChecked(offset, castagnoliTable)
+func NewSymbols(factory *streamencoding.DecbufFactory, version, offset int, doChecksum bool) (s *Symbols, err error) {
+	tab := castagnoliTable
+	if !doChecksum {
+		tab = nil
+	}
+
+	d := factory.NewDecbufAtChecked(offset, tab)
```
Comment (pracucci) [inline pkg/storegateway/indexheader/index/symbols.go:50]: [nit] Even if the logic is the same, calling `NewDecbufAtUnchecked()` when `!doChecksum` may be easier to read. An alternative with nearly the same lines of code could be:

```
var d Decbuf

if doChecksum {
  d = factory.NewDecbufAtChecked(offset, castagnoliTable)
} else {
  d = factory.NewDecbufAtUnchecked(offset)
}
```

File: pkg/storegateway/indexheader/index/postings.go [hunk 4]
```diff
@@ -51,11 +51,15 @@ type PostingOffsetTableV1 struct {
...
+func NewPostingOffsetTable(factory *streamencoding.DecbufFactory, tableOffset int, indexVersion int, indexLastPostingListEndBound uint64, postingOffsetsInMemSampling int, doChecksum bool) (PostingOffsetTable, error) {
 	if indexVersion == index.FormatV1 {
 		return newV1PostingOffsetTable(factory, tableOffset, indexLastPostingListEndBound)
 	} else if indexVersion == index.FormatV2 {
-		return newV2PostingOffsetTable(factory, tableOffset, indexLastPostingListEndBound, postingOffsetsInMemSampling)
+		tab := castagnoliTable
+		if !doChecksum {
+			tab = nil
+		}
+		return newV2PostingOffsetTable(factory, tab, tableOffset, indexLastPostingListEndBound, postingOffsetsInMemSampling)
```
Comment (pracucci) [inline pkg/storegateway/indexheader/index/postings.go:62]: [nit] Have you considered push down `doChecksum` to `newV2PostingOffsetTable()`?

File: CHANGELOG.md (hunk 1 above)
Comment (dimitarvdimitrov) [reply CHANGELOG.md:7]: I copied all the sections from the previous release and added them to the unreleased section in a69eed43afe5ae44378259d514f65136f2d93203

**Pull request merged**